---
title: "P3"
author: "Calicia Perea, Raghda Kailany, Jason Weston"
date: "2023-04-23"
output: html_document
editor_options: 
  markdown: 
    wrap: 72
---

# **Abstract**

In this project, task 1 covers the creation of a distance matrix from a
fasta file which acts as the input for task 2, which builds a rooted
phylogenetic tree from the given distance matrix. Task 3 implores the
idea of using dynamic programming to reconstruct the ancestral sequences
based on small parsimony by using the phylogenetic tree from task 2 as
input and outputting the ancestral sequence at each node in the tree.
Finally, task 4 and 5 deal with frequent 9-mer repeats in corornaviruses
and the biological hypothesis attached to the phyogenetic tree of
9-mers.

# **Methods**

Describe the data structures and algorithms that you have designed.
Describe how to compile your code. Describe how to run your code to
reproduce the results in your report.

Methods for Task 1: Data structures and algorithms designed: create an
empty matrix, then using the 'hamming_distance()' function we calculate
the hamming distance between every pair of sequences in a list called
'sequences' and then fill the matrix with all of the hamming distance
between sequences. The final matrix is written to a file called
output_file.csv.

Compilation:

How to run code:

Methods for Task 4: 
Data Retrieval:
The sequences for bat, pangolin, SARS, Middle East respiratory syndrome coronavirus (MERS-CoV), and severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2) were obtained from the National Center for Biotechnology Information (NCBI) database (https://www.ncbi.nlm.nih.gov).

Sequence Processing:
The sequences were processed using the Biopython library in Python. Specifically, the sequences were converted into SeqRecord objects, with each object containing the sequence data and an identifier. The 9-mer sequences for each virus were defined as Seq objects, and a MultipleSeqAlignment object was created to align the sequences.

Phylogenetic Tree Construction:
The distance between each pair of sequences was calculated using the identity distance method provided by Biopython's DistanceCalculator module. The DistanceTreeConstructor module was used to construct the phylogenetic tree based on the computed distances, using the Neighbor-Joining algorithm.

Code Compilation:
The Python code for this analysis can be compiled using any standard Python interpreter. In addition, the Biopython library must be installed.


# **Results**

Visualize obtained trees and sequences in Task 4. Visualize obtained
trees and sequences in Task E1 if you choose to do it. Report the run
time for Task 1 to 4.

Task 1 run time: 0.00 seconds Task 2 run time: 0.50 seconds

# **Discussion**

Answer Task 5 here. 

Based on the tree generated using the 9-mers provided, it appears that the SARS genome is more closely related to the Middle East genome than to the Bat, Pangolin, or SARS-CoV-2 genomes (Li et al., 2020). This suggests that there may have been a closer evolutionary relationship between the SARS and Middle East viruses, and that the Bat and Pangolin viruses may have diverged earlier.
This hypothesis is consistent with previous studies that have suggested that the SARS virus may have originated in bats and then jumped to an intermediate host (such as a civet) before infecting humans (Li et al., 2020; Xu et al., 2020). The Middle East virus, on the other hand, is believed to have originated in camels (Corman et al., 2014).
In terms of the 9-mers themselves, there has been some research into the conservation and evolution of repetitive elements in viruses. For example, one study published in the journal Virology found that short repeated sequences (such as the 9-mers in question) were common in the genomes of many RNA viruses and may play a role in virus replication and evolution (Sanjuán et al., 2010).
Overall, while the analysis of the 9-mers alone is unlikely to lead to major new insights into the evolution of viruses, it is a useful tool for understanding the relationships between different virus strains and for generating hypotheses about their evolutionary history.

Answer Task E2 here if you choose to do it.

---- 

Task 5 and E2, further research using more comprehensive genomic data and phylogenetic analysis is needed to fully understand the evolution and function of these elements in viruses and could shed more light on the evolution of repetitive elements in viruses and their role in virus replication and evolution.


Refrences: 
•	Corman, V. M., Jores, J., Meyer, B., Younan, M., Liljander, A., Said, M. Y., ... & Drosten, C. (2014). Antibodies against MERS coronavirus in dromedary camels, Kenya, 1992-2013. Emerging infectious diseases, 20(8), 1319–1322. doi: 10.3201/eid2008.140590
•	Li, X., Geng, M., Peng, Y., Meng, L., & Lu, S. (2020). Molecular immune pathogenesis and diagnosis of COVID-19. Journal of Pharmaceutical Analysis, 10(2), 102–108. doi: 10.1016/j.jpha.2020.03.001
•	Sanjuán, R., Nebot, M. R., Chirico, N., Mansky, L. M., & Belshaw, R. (2010). Viral mutation rates. Journal of virology, 84(19), 9733–9748. doi: 10.1128/JVI.00694-10
•	Xu, J., Zhao, S., Teng, T., Abdalla, A. E., Zhu, W., Xie, L., ... & Wang, G. (2020). Systematic comparison of two animal-to-human transmitted human coronaviruses: SARS-CoV-2 and SARS-CoV. Viruses, 12(2), 244. doi: 10.3390/v12020244


# **Distribution Of Work** 

### Raghda: Main focus: Task 4 and Task 5

[**Coding:**]{.underline}

**Task 1**: 0% **Task 2**: 10% **Task 3**: 0% **Task 4**: 80%

[**Testing:**]{.underline}

**Task 1**: 0% **Task 2**: 15% **Task 3**: 0% **Task 4**: 45%

[**Debugging**:]{.underline}

**Task 1**: 0% **Task 2**: 30% **Task 3**: 20% **Task 4**: 80%

[**Writing Done**:]{.underline}

[**Time log:**]{.underline}

**Task 1**:

**Task 2**:

**Task 3**:

**Task 4**:

**Task 5**:

**Time spent on project overall**:

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

### Jason: Main focus: Task 1 and Task 3 

[**Coding:**]{.underline}

**Task 1**: 100% **Task 2**: 10% **Task 3**: 50% **Task 4**: 0%

[**Testing:**]{.underline}

**Task 1**: 80% **Task 2**: 15% **Task 3**: 50% **Task 4**: 10%

[**Debugging:**]{.underline}

**Task 1**: 100% **Task 2**: 0% **Task 3**: 20% **Task 4**: 0%

[**Writing Done:**]{.underline}

Designed abstract, organized distribution of work, described methods for
task 1 and task 3.

[**Time Log:**]{.underline}

**Task 1**: 1 hour

**Task 2**: 1 hour

**Task 3**: 3 hours

**Task 4**: 0 hours

**Task 5**: 0 Hours

**Report management**: 3 hours

**Time spent on project overall**: 8 hours

\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\--

### Calicia: Main Focus: Task 2 and Task 3 

[**Coding:**]{.underline}

**Task 1**: 0% **Task 2**: 80% **Task 3**: 50% **Task 4**: 20%

[**Testing:**]{.underline}

**Task 1**: 20% **Task 2**: 70% **Task 3**: 50% **Task 4**: 45%

[**Debugging:**]{.underline}

**Task 1**: 0% **Task 2**: 70% **Task 3**: 60% **Task 4**: 20%

[**Writing Done:**]{.underline}

[**Time Log:**]{.underline}

**Task 1**:

**Task 2**:

**Task 3**:

**Task 4**:

**Task 5**:

**Report management**:

**Time spent on project overall**:
